### PR TITLE
swarm: add early muxer selection to swarm metrics

### PR DESCRIPTION
**Pull request by sukunrt**
fixes: https://github.com/libp2p/go-libp2p/issues/2021

### CODE REVIEW

**Comment by p-shahi** [comment]
Please update the Changelog.md file in this section with these changes https://github.com/libp2p/go-libp2p/blob/master/CHANGELOG.md#additional-metrics-

File: core/network/conn.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ type ConnectionState struct {
 	Security protocol.ID
 	// the transport used on this connection. For example: tcp
 	Transport string
+	// indicates whether StreamMultiplexer was selected using inlined muxer negotiation
+	EarlyMuxerSelection bool
```
Comment (marten-seemann) [inline core/network/conn.go:47]:
```suggestion
	UsedEarlyMuxerNegotiation bool
```

File: p2p/net/upgrader/conn.go [hunk 2]
```diff
@@ -16,8 +16,9 @@ type transportConn struct {
 	scope     network.ConnManagementScope
 	stat      network.ConnStats
 
-	muxer    protocol.ID
-	security protocol.ID
+	muxer               protocol.ID
+	security            protocol.ID
+	earlyMuxerSelection bool
```
Comment (marten-seemann) [inline p2p/net/upgrader/conn.go:21]:
```suggestion
	usedEarlyMuxerNegotiation bool
```

File: dashboards/swarm/swarm.json [hunk 3]
```diff
@@ -1059,6 +1059,139 @@
       "title": "Connection Duration",
       "type": "timeseries"
     },
+    {
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply dashboards/swarm/swarm.json:1062]: made it a pie chart

File: core/network/conn.go (hunk 1 above)
Comment (sukunrt) [reply core/network/conn.go]: fixed

File: p2p/net/upgrader/conn.go (hunk 2 above)
Comment (sukunrt) [reply p2p/net/upgrader/conn.go]: fixed.

File: dashboards/swarm/swarm.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (marten-seemann) [reply dashboards/swarm/swarm.json:1062]: Looks good!

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**